pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -296,15 +298,290 @@ func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, err
+			return
+		}
+	} else {
+		// Otherwise kill any running containers in this pod which are not to keep
+		runningContainerStatues := podStatus.GetRunningContainerStatuses()
+		for _, containerStatus := range runningContainerStatues {
+			if _, keep := podContainerChanges.ContainersToKeep[containerStatus.ID]; !keep {
+				glog.V(3).Infof("Killing unwanted container %q(id=%q) for pod %q", containerStatus.Name, containerStatus.ID, format.Pod(pod))
+
+				var podContainer *api.Container
+				var killMessage string
+				for i, c := range pod.Spec.Containers {
+					if c.Name == containerStatus.Name {
+						podContainer = &pod.Spec.Containers[i]
+						killMessage = podContainerChanges.ContainersToStart[i]
+						break
+					}
+				}
+				killContainerResult := kubecontainer.NewSyncResult(kubecontainer.KillContainer, containerStatus.Name)
+				result.AddSyncResult(killContainerResult)
+				if err := m.killContainer(pod, containerStatus.ID, podContainer, killMessage, nil); err != nil {
+					killContainerResult.Fail(kubecontainer.ErrKillContainer, err.Error())
+					glog.Errorf("killContainer %q(id=%q) for pod %q failed: %v", containerStatus.Name, containerStatus.ID, format.Pod(pod), err)
+					return
+				}
+			}
+		}
+	}
+
+	// We pass the value of the podIP down to generatePodSandboxConfig and
+	// generateContainerConfig, which in turn passes it to various other
+	// functions, in order to facilitate functionality that requires this
+	// value (hosts file and downward API) and avoid races determining
+	// the pod IP in cases where a container requires restart but the
+	// podIP isn't in the status manager yet.
+	//
+	// We default to the IP in the passed-in pod status, and overwrite it if the
+	// sandbox needs to be (re)started.
+	podIP := ""
+	if podStatus != nil {
+		podIP = podStatus.IP
+	}
+
+	// Create a sandbox for the pod.
+	podSandboxID := podContainerChanges.SandboxID
+	if podContainerChanges.CreateSandbox && len(podContainerChanges.ContainersToStart) > 0 {
+		var msg string
+		var err error
+
+		glog.V(4).Infof("Creating sandbox for pod %q", format.Pod(pod))
+		createSandboxResult := kubecontainer.NewSyncResult(kubecontainer.CreatePodSandbox, format.Pod(pod))
+		result.AddSyncResult(createSandboxResult)
+		podSandboxID, msg, err = m.createPodSandbox(pod, podIP, podContainerChanges.Attempt+1)
+		if err != nil {
+			createSandboxResult.Fail(kubecontainer.ErrCreatePodSandbox, msg)
+			glog.Errorf("createPodSandbox for pod %q failed: %v", format.Pod(pod), err)
+			return
+		}
+
+		setupNetworkResult := kubecontainer.NewSyncResult(kubecontainer.SetupNetwork, podSandboxID)
+		result.AddSyncResult(setupNetworkResult)
+		if !kubecontainer.IsHostNetworkPod(pod) {
+			glog.V(3).Infof("Calling network plugin %s to setup pod for %s", m.networkPlugin.Name(), format.Pod(pod))
+			// Setup pod network plugin with sandbox id
+			err = m.networkPlugin.SetUpPod(pod.Namespace, pod.Name, kubecontainer.ContainerID{
```

Comment:
Already added the TODO to NetworkPlugin interface, see https://github.com/kubernetes/kubernetes/pull/31847/files#diff-6d531dbca1d7006455f70f94a31c4829R71.